### PR TITLE
Implementation of phone question type (with US and CA support)

### PR DESCRIPTION
**Pull request by swatkat1**
### Description
This is the Second iteration of the Phone Type question. 
This PR contains:
1. The Admin UI changes 
2. The Applicant changes
3. Unit tests to cover both

This PR doesn't include any Browser tests. It will be a fast followup to this PR. 
## Release notes

This is the Second iteration of the Phone Type question. With the flag- phone_question_type_enabled set to true, the phone type question can be tested in staging.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes # https://github.com/civiform/civiform/issues/452

### CODE REVIEW

File: server/app/assets/javascripts/phone.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import Cleave from 'cleave.js'
+
+export function init() {
+  const element = document.getElementById('cf-phone-number')
+  const countryElement = document.getElementById('cf-phone-country-code')
+  if (element && countryElement) {
+    new Cleave(element, {
+      numericOnly: true,
+      blocks: [0, 3, 3, 4],
+      delimiters: ['(', ') ', '-'],
```
Comment (bion) [inline server/app/assets/javascripts/phone.ts:10]: **nitpick (blocking):** add a comment specifying what this config does, particularly what the `blocks` attribute does

File: server/app/services/CfJsonDocumentContext.java [hunk 2]
```diff
@@ -89,6 +91,21 @@ public boolean hasValueAtPath(Path path) {
...
   }
 
+  public void putPhoneNumber(Path path, String phoneNumber) {
+    if (phoneNumber.isEmpty()) {
+      putNull(path);
+      return;
+    }
+    // Currently on US and CA numbers are supported, hence testing for just the two regions.
+    // for future cases, it is better to get the countrycode and test against the country code, not
+    // just all possible regions.
```
Comment (bion) [inline server/app/services/CfJsonDocumentContext.java:101]: **nitpick (blocking):** 
```suggestion
    // Currently only US and CA numbers are supported, hence testing for just the two regions.
    // For future cases, it is better to get the countrycode and test against the country code, not
    // just all possible regions.
```

File: server/app/services/CfJsonDocumentContext.java [hunk 3]
```diff
@@ -89,6 +91,21 @@ public boolean hasValueAtPath(Path path) {
...
+    if (PHONE_NUMBER_UTIL.isPossibleNumber(phoneNumber, "US")
+        || PHONE_NUMBER_UTIL.isPossibleNumber(phoneNumber, "CA")) {
+      put(path, phoneNumber.replaceAll("[^0-9]", ""));
+    } else {
+      throw new IllegalArgumentException("invalid_phone_input");
```
Comment (bion) [inline server/app/services/CfJsonDocumentContext.java:106]: **suggestion (blocking):** include more info in the exception e.g. "Invalid phone number format", consider also including the format in the exception with the numbers redacted e.g. `String.format("Invalid phone number format: %s", phoneNumber.replaceAll("\\d", "X"))`

File: server/app/services/export/JsonExporter.java [hunk 4]
```diff
@@ -206,4 +228,13 @@ private DocumentContext makeEmptyJsonArray() {
   private DocumentContext makeEmptyJsonObject() {
     return JsonPathProvider.getJsonPath().parse("{}");
   }
+
+  public String getFormattedPhoneNumber(String phoneNumberValue, String countryCode) {
+    try {
+      Phonenumber.PhoneNumber phoneNumber = PHONE_NUMBER_UTIL.parse(phoneNumberValue, countryCode);
+      return PHONE_NUMBER_UTIL.format(phoneNumber, PhoneNumberUtil.PhoneNumberFormat.INTERNATIONAL);
```
Comment (bion) [inline server/app/services/export/JsonExporter.java:235]: **question (blocking):** what standard does this formatting follow? will need to provide it in our [API docs](https://docs.civiform.us/it-manual/api/program-applications) (see table of "Question type export values").

Add javadoc for this method that mentions the formatting standard.

File: server/app/services/export/JsonExporter.java [hunk 5]
```diff
@@ -206,4 +228,13 @@ private DocumentContext makeEmptyJsonArray() {
   private DocumentContext makeEmptyJsonObject() {
     return JsonPathProvider.getJsonPath().parse("{}");
   }
+
+  public String getFormattedPhoneNumber(String phoneNumberValue, String countryCode) {
```
Comment (bion) [inline server/app/services/export/JsonExporter.java:232]: **suggestion (blocking):** methods show have the most restrictive visibility setting they can. Make this method `private` or if you're testing directly make it package-private (remove the visibility keyword) and add the `@VisibleForTesting` annotation

File: server/app/views/admin/programs/ProgramBlockEditView.java [hunk 6]
```diff
@@ -1,10 +1,7 @@
 package views.admin.programs;
 
 import static com.google.common.base.Preconditions.checkNotNull;
-import static featureflags.FeatureFlag.ESRI_ADDRESS_CORRECTION_ENABLED;
-import static featureflags.FeatureFlag.INTAKE_FORM_ENABLED;
-import static featureflags.FeatureFlag.NONGATED_ELIGIBILITY_ENABLED;
-import static featureflags.FeatureFlag.PROGRAM_ELIGIBILITY_CONDITIONS_ENABLED;
+import static featureflags.FeatureFlag.*;
```
Comment (bion) [inline server/app/views/admin/programs/ProgramBlockEditView.java:4]: **nitpick:** revert this change https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: server/app/views/admin/questions/QuestionConfig.java [hunk 7]
```diff
@@ -5,13 +5,7 @@
...
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
-import forms.AddressQuestionForm;
-import forms.EnumeratorQuestionForm;
-import forms.IdQuestionForm;
-import forms.MultiOptionQuestionForm;
-import forms.NumberQuestionForm;
-import forms.QuestionForm;
-import forms.TextQuestionForm;
+import forms.*;
```
Comment (bion) [inline server/app/views/admin/questions/QuestionConfig.java:8]: **nitpick:** revert this change

File: server/app/views/admin/questions/QuestionConfig.java [hunk 8]
```diff
@@ -87,6 +83,15 @@ public static Optional<DivTag> buildQuestionConfig(QuestionForm questionForm, Me
     }
   }
 
+  private QuestionConfig addPhoneConfig() {
+    content.with(
+        new DivTag()
+            .withText(
+                "This supports only US and CA phone numbers. If you need other international"
+                    + " numbers, Please use Text type questions"));
```
Comment (bion) [inline server/app/views/admin/questions/QuestionConfig.java:91]: **suggestion:** 
```suggestion
                "This supports only US and CA phone numbers. If you need other international"
                    + " numbers, please use a Text question."));
```

File: server/app/views/admin/questions/QuestionPreview.java [hunk 9]
```diff
@@ -46,7 +46,7 @@ public static DivTag renderQuestionPreview(
     try {
       renderedQuestion = buildQuestionRenderer(type, messages, fileUploadViewStrategy);
     } catch (UnsupportedQuestionTypeException e) {
-      renderedQuestion = div().withText(e.toString());
+      throw new RuntimeException(e);
```
Comment (bion) [inline server/app/views/admin/questions/QuestionPreview.java:49]: **suggestion (blocking):** This looks unrelated to implementing phone number question type. Probably better to do in a separate PR. What's motivating this change?

File: server/app/views/admin/questions/QuestionTranslationView.java [hunk 10]
```diff
@@ -112,6 +112,7 @@ private Optional<DomContent> getQuestionTypeSpecificContent(
       case NAME: // fallthrough intended
       case NUMBER: // fallthrough intended
       case TEXT: // fallthrough intended
+      case PHONE:
```
Comment (bion) [inline server/app/views/admin/questions/QuestionTranslationView.java:115]: **praise:** 
```suggestion
      case PHONE:  // fallthrough intended
```

File: server/app/views/components/CreateQuestionButton.java [hunk 11]
```diff
@@ -48,6 +49,9 @@ public static DivTag renderCreateQuestionButton(
                 "hidden");
 
     for (QuestionType type : QuestionType.values()) {
+      if (!phoneQuestionTypeEnabled && type == QuestionType.PHONE) {
```
Comment (bion) [inline server/app/views/components/CreateQuestionButton.java:52]: **nitpick:** use `equals` for comparing enums
```suggestion
      if (!phoneQuestionTypeEnabled && QuestionType.PHONE.equals(type)) {
```

File: server/app/views/questiontypes/PhoneQuestionRenderer.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  protected DivTag renderInputTag(
+      ApplicantQuestionRendererParams params,
+      ImmutableMap<Path, ImmutableSet<ValidationErrorMessage>> validationErrors,
+      ImmutableList<String> ariaDescribedByIds) {
+    PhoneQuestion phoneQuestion = question.createPhoneQuestion();
+
+    Messages messages = params.messages();
+
+    SelectWithLabel countryCodeField =
+        (SelectWithLabel)
```
Comment (bion) [inline server/app/views/questiontypes/PhoneQuestionRenderer.java:39]: **nitpick:** 
```suggestion
```

File: server/app/views/questiontypes/PhoneQuestionRenderer.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    return div()
+        .withClasses("grid")
+        .with(countryCodeField.getSelectTag(), phoneField.getInputTag());
+  }
+
+  private static ImmutableList<SelectWithLabel.OptionValue> countryOptions() {
+    return ImmutableList.of(
+        SelectWithLabel.OptionValue.builder().setLabel("United States").setValue("US").build(),
+        SelectWithLabel.OptionValue.builder().setLabel("Canada").setValue("CA").build());
+  }
```
Comment (bion) [inline server/app/views/questiontypes/PhoneQuestionRenderer.java:85]: **suggestion (blocking):** any reason for this to be a static method and not a constant?
```suggestion
  private static final ImmutableList<SelectWithLabel.OptionValue> COUNTRY_OPTIONS = ImmutableList.of(
        SelectWithLabel.OptionValue.builder().setLabel("United States").setValue("US").build(),
        SelectWithLabel.OptionValue.builder().setLabel("Canada").setValue("CA").build());
```

File: server/conf/messages.en-US [hunk 14]
```diff
@@ -166,3 +166,4 @@ validation.textTooLong=Must contain at most {0} characters.
 validation.textTooShort=Must contain at least {0} characters.
 validation.tooFewSelections=Please select at least {0}.
 validation.tooManySelections=Please select fewer than {0}.
+
```
Comment (bion) [inline server/conf/messages.en-US:169]: **suggestion (blocking):** add the english messages here as well

**Review by bion — CHANGES_REQUESTED** [review]

File: server/conf/messages.en-US (hunk 14 above)
Comment (swatkat1) [reply server/conf/messages.en-US:169]: If I add it here, the conf/Messagestests.java are failing. If anything is added to the ignore list, the tests expect that it only be present in conf/messages and not in any other translation files.

File: server/app/views/admin/programs/ProgramBlockEditView.java (hunk 6 above)
Comment (swatkat1) [reply server/app/views/admin/programs/ProgramBlockEditView.java:4]: This is the second time its happening. I am wondering if the bin/fmt is doing it or intelliJ

File: server/app/views/admin/questions/QuestionConfig.java (hunk 7 above)
Comment (swatkat1) [reply server/app/views/admin/questions/QuestionConfig.java:8]: This is the second time its happening. I am wondering if the bin/fmt is doing it or intelliJ

File: server/app/views/admin/questions/QuestionPreview.java (hunk 9 above)
Comment (swatkat1) [reply server/app/views/admin/questions/QuestionPreview.java:49]: This was the issue we discovered during the pair-programming session where I couldn't clearly see why the PhoneQuestion wasn't visible. You suggested to make this change to make it more obvious rather than silently failing. Since it was discovered as part of this PR, I left it here.

File: server/app/views/questiontypes/PhoneQuestionRenderer.java (hunk 12 above)
Comment (swatkat1) [reply server/app/views/questiontypes/PhoneQuestionRenderer.java]: Can't remove this unfortunately. I see this error: 
Required type: SelectWithLabel
Provided: FieldWithLabel

File: server/app/assets/javascripts/phone.ts (hunk 1 above)
Comment (swatkat1) [reply server/app/assets/javascripts/phone.ts:10]: done.

File: server/app/services/CfJsonDocumentContext.java (hunk 3 above)
Comment (swatkat1) [reply server/app/services/CfJsonDocumentContext.java:106]: Thanks Bion. Actually this not only catches formatting error but also invalid number like 555-###-####, though they look legitimate, they are not actual phone numbers. I have updated the comment in this method.

File: server/app/services/export/JsonExporter.java (hunk 5 above)
Comment (swatkat1) [reply server/app/services/export/JsonExporter.java:232]: Done!

File: server/app/views/admin/programs/ProgramBlockEditView.java (hunk 6 above)
Comment (bion) [reply server/app/views/admin/programs/ProgramBlockEditView.java]: IntelliJ does it by default, you can disable it: https://intellij-support.jetbrains.com/hc/en-us/community/posts/206203659-Turn-off-Wildcard-imports-

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/services/export/JsonExporter.java (hunk 4 above)
Comment (swatkat1) [reply server/app/services/export/JsonExporter.java:240]: Thanks for catching this. I have changed it to E164 format.

**Review by bion — APPROVED** [review]
